include/robot_interfaces/opencv_driver.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <ctime>
+#include <iostream>
+
+#include <real_time_tools/process_manager.hpp>
+#include <real_time_tools/thread.hpp>
+#include <real_time_tools/threadsafe/threadsafe_timeseries.hpp>
+#include <real_time_tools/timer.hpp>
+#include "opencv2/opencv.hpp"
+
+namespace robot_interfaces
+{
+/**
+ * @brief Driver for interacting with any camera using OpenCV.
+ */
+
+template <typename OpenCVObservation>
+class OpenCVDriver
+{
+public:
+    cv::VideoCapture video_capture;
+    real_time_tools::Timer timer;
+
+    OpenCVDriver()
+    {
+        cv::VideoCapture cap(0);
+        video_capture = cap;
+    }
+
+    /**
+     * @brief Find out if the camera can be accessed and
+     * if the video capture has been started.
+     */
+
+    int is_grabbing_successful()
+    {
+        if (!video_capture.isOpened())
+        {
+#ifdef VERBOSE
+            std::cout << "Could not access camera stream :(" << std::endl;
+#endif
+            return -1;
+        }
+        else
+        {
+#ifdef VERBOSE
+            std::cout << "Succeeded in accessing camera stream!" << std::endl;
+#endif
+            return 1;
+        }
+    }
+
+    /**
+     * @brief Grab frames along with their timestamps one by one.
+     *
+     * @return Image frame consisting of an image matrix and the time at
+     * which it was grabbed.
+     */
+
+    OpenCVObservation grab_frame()
+    {
+        OpenCVObservation image_frame;
+        cv::Mat frame;
+        long double current_time = timer.get_current_time_sec();
```

Comment:
`get_current_time_sec` is a static method so you don't need an actual `Timer` object.  Instead you can call it like this: `real_time_tools::Timer::get_current_time_sec()`.

Since capturing the frame may take some time (e.g. because we need to wait for the camera to actually record the next frame), it would be better to get the timestamp afterwards, not before.

Further this function returns a `double`, not a `long double`.

Reply:
Okay.

But the timestamp is a `long double`.  I had defined the time_stamp field of the observation to be a `long double Timstamp`. Should I change that?

Reply:
Ok, understood. I had mistaken the type to be the same for the TimeStamp from `time_series` and from `real_time_tools`.